[Inductor] Async Pipelined Autotune

PaulZhang12 opened the pull request:
Async-Pipelined Autotuning is a path for Inductor max-autotune-gemm to do autotuning in a subprocess during lowering time, overlapping autotuning with lowering/scheduling time in parallel to reduce compilation overhead. Below, we highlight the base case versus this new path:

**Baseline**
1. Go through lowering path for each gemm, calling precompile synchronously for each choice and returning
2. During scheduling time, `get_min_choice` for each MultiTemplateBuffer does the actual autotuning synchronously. 

However, this approach results in the GPU being idle up until the autotuning time for each choice is actually needed. Async-Pipelined Autotuning addresses this issue by being able to kick off autotuning for each choice async as soon as possible, to attempt to keep the GPU fed throughout the compilation process for max-autotune-gemm.

**Async-Pipelined Autotuning**
1. Compile on main process and thread, precompilation thread pool on main process, 1 autotuning subprocess
2. Precompilation is now truly async, for each gemm that is lowered, do kick off "starting precompilation" completely async in a separate thread pool and return immediately
3. Kick off autotuning in subprocess for extern choices before precompilation of triton choices, want to keep GPU subprocess fed
4. After precompilation is done, separate thread invokes the autotune subprocess to tune that precompiled choice


Below is a diagram of the general architecture with Diode, the ML GEMM model that can be plugged into Inductor:

<img width="4733" height="5371" alt="image" src="https://github.com/user-attachments/assets/70fd83c6-e629-4a35-a08c-9e52647c16b4" />


**E2E numbers**
Based on a portion of a representative internal model:
Base: ~14s compilation time for graph
Topk=1 Diode + disable epilogue fusion: ~30s compilation time
Topk=1 Diode + disable epilogue fusion + Async lower: ~27s compilation time

**TODO**

* Support Prologue/Epilogue Fusion
* Support Subgraph Choice Callers



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
+    import torch
+
+    # Initialize dummy tensor for CUDA context
+    if torch.cuda.is_available():
```
mostly out of curiosity: Given our discussion in chat, I wouldn't think this is necessary. We believe that importing torch is enough to init CUDA?

PaulZhang12 replied:
Will test!

masnesral commented on torch/_inductor/config.py:
```diff
@@ -466,6 +466,11 @@ def prologue_fusion_enabled() -> bool:
 )
 
+# Pipeline autotuning for max-autotune-gemm. Overlap lowering and benchmarking on GPU
+pipeline_max_autotune_gemm = (
```
To make internal rollout easier, you probably want to create a JK. Here's an example of how we do that elsewhere in this file: https://fburl.com/code/y8ez9f7e

PaulZhang12 replied:
@masnesral yeah, this is enabled off by default for now and I intend to enable it on a case by case basis, no wide rollout at all currently

masnesral replied:
I meant if you just create the JK now, it provides you a convenient way to do that rollout.

masnesral commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2830,40 +2831,93 @@ def __call__(
+                precompile_instance = PrecompileThreadPool.get_instance(
+                    num_workers=get_num_workers()
+                )
+                precompile_future = precompile_instance.submit(
```
Also left a msg in chat. I can't quite figure out what's going on here. The existing code is already so confusing :/  Is this thread pool actually being used to precompile anything? I _think_ the precompile_fn at this point is just a function to wait on all precompile futures submitted to the async_compile pool. This "precompile_fn" is just waiting for the futures to finish.

PaulZhang12 replied:
Responded with this in chat:
"This PR basically makes it so that the do_autotuning is wrapped in a future completely. That way, when the precompilation is complete for a single triton choice, another thread will send it to the autotune subprocess as soon as possible"

PaulZhang12 commented on torch/_inductor/autotune_process.py:
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
+        for choice in choices:
+            choice_hash = AsyncAutotuner.get_choice_hash(choice, inputs_key)
+            timings[choice] = AsyncAutotuner.choice_hash_to_future[choice_hash].result(
+                timeout=timeout
```
Actually I think this is missing some code here, which is basic caching logic (not reautotuning on choices that have been autotuned before). Will add that back in, but this allows us to nicely store the result for future autotune choices

nmacchioni commented on torch/_inductor/autotune_process.py:
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
+            with cls._lock:
+                if cls._instance is None:
+                    cls._instance = cls(num_workers)
+        return cls._instance
```
Copy-pasting my comment from the last PR since it is still relevant:

This doesn't seem right, originally each call to precompile would get its own thread pool with 1 thread per 1 choice (assuming max # threads is larger). Now each call to precompile shares the same thread pool, which has # of threads = to the # of choices on the first call to precompile? So not only do subsequent precompile calls have to wait for prior precompile calls to free up threads, but we might never have enough threads to process an entire precompile at once (if the first call has less choices than subsequent calls)

PaulZhang12 replied:
I see ~5% faster compile time from not constructing a threadpool for every choice

PaulZhang12 replied:
Threadpool per shape -> single thread for every choice. Basically the default. Reconstructing this thread pool in another thread pool generally seems bad for perf

nmacchioni replied:
Sure, I agree that creating a new threadpool per call is not ideal. I'm just saying that the logic for setting the max_workers is not consistent with the prior implementation.

Previously, one threadpool per shape meant one thread for every choice (in the common case).

Now, one threadpool for all shapes has some # of max_workers dependent on who initialized the threadpool. So now we'll only have ~20 (avg # of configs) threads floating around for all of the jobs in total. So let's say instead of top-1 pruning you used top-10 or something. The threadpool would be initialized with 11 (10 triton + 1 aten) threads at the first call. Any subsequent calls would submit jobs to this threadpool, but would only be queued since the 11 threads are already in use. So you'd end up getting a backup.

nmacchioni reviewed (commented):
What benefit do you see in having PrecompileThreadPool rather than a regular ThreadPoolExecutor in the init of AlgorithmSelectorCache?

PaulZhang12 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2830,40 +2831,93 @@ def __call__(
-                    for c in choices
-                    if not hasattr(c, "hint_override")
-                    or c.hint_override == hint_override
+            if config.pipeline_max_autotune_gemm:
```
I'm not sure this will actually be better... we depend on quite a few local variables and we would just be passing a lot of them around anyways

eellison reviewed (approved):
looks good - mostly deferring to @masnesral

PaulZhang12 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -3394,8 +3455,6 @@ def wait_on_futures():
             if exceptions:
                 _log_autotune_exceptions(exceptions)
 
-            executor.shutdown(wait=True)
```
Yes that's right, I forgot to push another change

PaulZhang12 committed:
Async Pipelined Autotune (#170128)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/170128

Test Plan: MaxAutotuneAsyncPipeline added which repeats most of the MaxAutotune tests but with this path

Reviewed By: eellison, masnesral

Differential Revision: D88889871

PaulZhang12 commented:
Got reverted in OSS while landed internally due to lint error. Redelegating to https://github.com/pytorch/pytorch/pull/170407
